File: LEETCODE/LEETCODE-PYTHON/2364.py
#2364. Count Number of Bad Pairs
#https://leetcode.com/problems/count-number-of-bad-pairs/description/

#Method-1 using brute force
class Solution:
    def countBadPairs(self, nums: List[int]) -> int:
        badPairs=0
        n=len(nums)

        for i in range(0,n-1):
            for j in range(i+1,n):
                if j-i != nums[j]-nums[i]:
                    badPairs+=1

        return badPairs
    
#Methos-2 using dict/map
class Solution:
    def countBadPairs(self, nums: List[int]) -> int:
        freq={}
        goodPairs=0
        n=len(nums)

        for i in range(0,n):
            temp=nums[i]-i
            goodPairs+=freq.get(temp,0)
            freq[temp]=freq.get(temp,0)+1

        return (n*(n-1))//2 - goodPairs

File: LEETCODE/LEETCODE-CPP/2364.cpp
//2364. Count Number of Bad Pairs
//https://leetcode.com/problems/count-number-of-bad-pairs/

//Method-1 brute force
class Solution {
public:
    long long countBadPairs(vector<int>& nums) {
        long long ans=0, n=nums.size();;

        for(long long i=0;i<n-1;i++){
            for(long long j=i+1;j<n;j++){
                if(j-i !=nums[j]-nums[i]){
                    ans++;
                }
            }
        }

        return ans;
    }
};


//Method-2 using map
class Solution {
    public:
        long long countBadPairs(vector<int>& nums) {
            long long goodPairs=0, n=nums.size();
            unordered_map<int,int> freq;
    
            for(long long i=0;i<n;i++){
                int temp=nums[i]-i;
                goodPairs+=freq[temp];
                freq[temp]++;
    
            }
    
            return (n*(n-1))/2 - goodPairs;
        }
    };